include/fiction/algorithms/physical_design/wiring_reduction.hpp

```diff
@@ -0,0 +1,935 @@
+#define FICTION_WIRING_REDUCTION_HPP
+
+#include "fiction/algorithms/path_finding/a_star.hpp"
+#include "fiction/algorithms/path_finding/cost.hpp"
+#include "fiction/algorithms/path_finding/distance.hpp"
+#include "fiction/layouts/bounding_box.hpp"
+#include "fiction/layouts/cartesian_layout.hpp"
+#include "fiction/layouts/clocking_scheme.hpp"
+#include "fiction/layouts/coordinates.hpp"
+#include "fiction/layouts/obstruction_layout.hpp"
+#include "fiction/traits.hpp"
+#include "fiction/utils/routing_utils.hpp"
+
+#include <mockturtle/traits.hpp>
+#include <mockturtle/utils/stopwatch.hpp>
+
+#include <cmath>
+#include <cstdint>
+#include <functional>
+#include <ostream>
+#include <utility>
+#include <vector>
+
+namespace fiction
+{
+
+/**
+ * This struct stores statistics about the wiring reduction process.
+ */
+struct wiring_reduction_stats
+{
+    /**
+     * Runtime of the wiring reduction process.
+     */
+    mockturtle::stopwatch<>::duration time_total{0};
+    /**
+     * Layout width before the wiring reduction process.
+     */
+    uint64_t x_size_before{0ull};
+    /**
+     * Layout height before the wiring reduction process.
+     */
+    uint64_t y_size_before{0ull};
+    /**
+     * Layout width after the wiring reduction process.
+     */
+    uint64_t x_size_after{0ull};
+    /**
+     * Layout height before the wiring reduction process.
+     */
+    uint64_t y_size_after{0ull};
+    /**
+     * Number of wire segments before the wiring reduction process.
+     */
+    uint64_t num_wires_before{0ull};
+    /**
+     * Number of wire segments after the wiring reduction process.
+     */
+    uint64_t num_wires_after{0ull};
+    /**
+     * Improvement in the number wire segments.
+     */
+    double wiring_improvement{0ull};
+    /**
+     * Improvement in layout area.
+     */
+    double area_improvement{0ull};
+
+    /**
+     * Reports the statistics to the given output stream.
+     *
+     * @param out Output stream.
+     */
+    void report(std::ostream& out = std::cout) const
+    {
+        out << fmt::format("[i] total time                      = {:.2f} secs\n", mockturtle::to_seconds(time_total));
+        out << fmt::format("[i] layout size before optimization = {} × {}\n", x_size_before, y_size_before);
+        out << fmt::format("[i] layout size after optimization  = {} × {}\n", x_size_after, y_size_after);
+        out << fmt::format("[i] area reduction                  = {}%\n", area_improvement);
+        out << fmt::format("[i] num. wires before optimization  = {}\n", num_wires_before);
+        out << fmt::format("[i] num. wires after optimization   = {}\n", num_wires_after);
+        out << fmt::format("[i] wiring reduction                = {}%\n", wiring_improvement);
+    }
+};
+
+namespace detail
+{
+
+/**
+ * The two search directions: horizontal (from left to right) and vertical (from top to bottom).
+ */
+enum class search_direction : uint8_t
+{
+    HORIZONTAL,
+    VERTICAL
```

Comment:
These two require individual docstrings, i.e.
```suggestion
    /**
     * ...
     */
    HORIZONTAL,
    /**
     * ...
     */
    VERTICAL
```